File: compiler/bench/partition-pivot.c
// expected: 499016000
// Partition around pivot benchmark - 1M iterations

#include <stdio.h>

#define SIZE 1000
#define ITERS 1000000

int arr[SIZE];

void swap_arr(int i, int j) {
    int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}

int do_partition(int pivot) {
    int i = 0, j = SIZE - 1;
    while (i < j) {
        while (i < j && arr[i] > pivot) i++;
        while (i < j && arr[j] <= pivot) j--;
        if (i < j) {
            swap_arr(i, j);
            i++;
            j--;
        }
    }
    return i;
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = (i * 17 + 23) % SIZE;
    }
}

long bench(void) {
    long sum = 0;
    for (int i = 0; i < ITERS; i++) {
        init_arr();
        sum += do_partition(i % SIZE);
    }
    return sum;
}

int main(void) {
    printf("%ld\n", bench());
    return 0;
}

File: compiler/bench/partition-pivot.fs
\ expected: 499016000
\ Partition around pivot benchmark - 1M iterations

1000 constant SIZE
1000000 constant ITERS
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: swap-arr ( i j -- )
  over arr@ over arr@
  rot arr! swap arr! ;

: do-partition ( pivot -- count-less )
  0 SIZE 1-               \ pivot i j
  begin 2dup < while
    begin 2dup < 3 pick 3 pick arr@ > and while
      1 under+
    repeat
    begin 2dup < 3 pick 2 pick arr@ <= and while
      1-
    repeat
    2dup < if 2dup swap-arr 1 under+ 1- then
  repeat
  nip nip ;

: init-arr ( -- )
  SIZE 0 do
    i 17 * 23 + SIZE mod i arr!
  loop ;

: bench ( -- sum )
  0
  ITERS 0 do
    init-arr
    i SIZE mod do-partition +
  loop ;

: main
  bench . cr ;
